### PR TITLE
Optional type documentation

### PR DESCRIPTION
**Pull request by neginraoof**
**Description**
- Adding the documentation about design and use-cases of Optional type
- Updating ONNX IR doc about ONNX types

### CODE REVIEW

**Comment by neginraoof** [comment]
@askhade @gramalingam @postrational 
Please let me know if you have comments.
Thanks!

File: docs/IR.md [hunk 1]
```diff
@@ -335,6 +335,7 @@ The following types are used to define the types of graph and node inputs and ou
 ONNX|dense tensors|Tensors are a generalization of vectors and matrices; whereas vectors have one dimension, and matrices two, tensors can have any number of dimensions, including zero. A zero-dimensional tensor is logically equivalent to a scalar value.
 ONNX|sequence|Sequences are dense, ordered, collections of elements that are of homogeneous types.
 ONNX|map|Maps are associative tables, defined by a key type and a value type.
+ONNX|optional|Optionals are a wrappers that may contain an element of tensor, sequence, or map type, or may be empty (containing none).
```
Comment (askhade) [inline docs/IR.md:338]: link the OnnxTypes.md here?

File: docs/ONNXTypes.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
+## Optional Type
+
+An optional type represents a reference to either an element (could be Tensor, Sequence, Map, or Sparse Tensor) or a null value. The optional type appears in model inputs, outputs, as well as intermediate values.  
+
+### Use-cases
+
+Optional type enables users to represent more dynamic typing senarios in ONNX. Similar to Optional[X] type hint in Python typing which is equivalent to Union[None, X], Optional typs in ONNX may reference a single element, or null.
```
Comment (askhade) [inline docs/ONNXTypes.md:7]: typo - Optional typs -> Optional types

File: docs/ONNXTypes.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+### Examples in PyTorch
+Optional type only appears in TorchScript graphs generated by jit script compiler. Scripting a model captures dynamic types where an optional value can be assigned either None or a value. 
+
+- Example 1
```
Comment (askhade) [inline docs/ONNXTypes.md:12]: for every example can you add how the resulting onnx graph would look like with this type added... I know we dont have operator support yet but you can add a placeholder for these ops

File: docs/IR.md [hunk 4]
```diff
@@ -309,7 +309,7 @@ More details can be found in [External Data](ExternalData.md).
 
 There are two official ONNX variants; the main distinction between the two is found in the supported types and the supported operators.
 
-With respect to supported types, the __ONNX__ definition recognizes only tensors as input and output types, while the Classical Machine Learning extension __ONNX-ML__ also recognizes sequences and maps.
+With respect to supported types, both __ONNX__ and __ONNX-ML__ definition recognize tensors, sparse tensors, sequences, maps, and optionals as input and output types.
```
Comment (gramalingam) [inline docs/IR.md:312]: I think it would be good to clarify IR-version distinctions: that is, optional type is supported from IR-version X, etc.

Comment (gramalingam) [reply docs/IR.md:312]: Even release version/IR-version etc.

File: docs/IR.md [hunk 5]
```diff
@@ -335,6 +335,7 @@ The following types are used to define the types of graph and node inputs and ou
 ONNX|dense tensors|Tensors are a generalization of vectors and matrices; whereas vectors have one dimension, and matrices two, tensors can have any number of dimensions, including zero. A zero-dimensional tensor is logically equivalent to a scalar value.
 ONNX|sequence|Sequences are dense, ordered, collections of elements that are of homogeneous types.
 ONNX|map|Maps are associative tables, defined by a key type and a value type.
+ONNX|optional|Optionals are a wrappers that may contain an element of tensor, sequence, or map type, or may be empty (containing none).
 
 ONNX currently does not define a sparse tensor type.
```
Comment (gramalingam) [inline docs/IR.md:340]: May need to update above line

File: onnx/defs/data_type_utils.cc [hunk 6]
```diff
@@ -190,6 +193,11 @@ void DataTypeUtils::FromString(
     return FromString(
         std::string(s.Data(), s.Size()),
         *type_proto.mutable_sequence_type()->mutable_elem_type());
+  } if (s.LStrip("optional")) {
```
Comment (gramalingam) [inline onnx/defs/data_type_utils.cc:196]: Is that supposed to be "else if"?

File: docs/IR.md (hunk 5 above)
Comment (gramalingam) [reply docs/IR.md:340]: Can we delete the above line "ONNX currently does not define a sparse tensor type." Thanks.

File: docs/ONNXTypes.md [hunk 7]
```diff
@@ -19,22 +19,39 @@ Optional type only appears in TorchScript graphs generated by jit script compile
...
                 -> (%12) 
                 block1(): 
                 -> (%x.1) 
-            return (%36) 
+            return (%6)
+
+    ONNX graph:
+
+        Graph(
+            %x.1 : Float(2, 3, strides=[3, 1], requires_grad=0, device=cpu),
```
Comment (gramalingam) [inline docs/ONNXTypes.md:42]: Can we drop all the extraneous attributes like strides, requires_grad, and device? They are somewhat confusing and misleading in an ONNX context. Same in all lines below as well. Thanks!

File: onnx/defs/schema.cc [hunk 8]
```diff
@@ -489,11 +499,13 @@ ATTR_SETTER_WITH_SINGLE_VALUE(float, f, AttributeProto::FLOAT)
 ATTR_SETTER_WITH_SINGLE_VALUE(std::string, s, AttributeProto::STRING)
 ATTR_SETTER_WITH_SINGLE_COMPLEXVALUE(TensorProto, t, AttributeProto::TENSOR)
 ATTR_SETTER_WITH_SINGLE_COMPLEXVALUE(GraphProto, g, AttributeProto::GRAPH)
+// ATTR_SETTER_WITH_SINGLE_COMPLEXVALUE(TypeProto, tp, AttributeProto::TYPE_PROTO)
```
Comment (askhade) [inline onnx/defs/schema.cc:502]: why is this commented?

**Review by gramalingam — APPROVED** [review]

File: onnx/defs/schema.cc (hunk 8 above)
Comment (neginraoof) [reply onnx/defs/schema.cc]: Thanks for catching this!

**Review by askhade — APPROVED** [review]

**Pull request merged**